hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java

```diff
@@ -1386,19 +1386,11 @@ private void addDatanode2ExistingPipeline() throws IOException {
        * Case 2: Failure in Streaming
        * - Append/Create:
        *    + transfer RBW
-       *
-       * Case 3: Failure in Close
-       * - Append/Create:
-       *    + no transfer, let NameNode replicates the block.
        */
     if (!isAppend && lastAckedSeqno < 0
         && stage == BlockConstructionStage.PIPELINE_SETUP_CREATE) {
       //no data have been written
       return;
-    } else if (stage == BlockConstructionStage.PIPELINE_CLOSE
```

Comment:
the stage is in PIPELINE_CLOSE state when when packet is the last in the block and all transferred data is acknowledged.

with this change, the re-replication which was triggered by NameNode block manager periodically, is moved to client side.

it looks fine to me. but HDFS re replication logic is complex. We should validate it with a test.

Reply:
Thank you for your suggestion. I just added a UT to this patch.